File: impls/erlang/src/step2_eval.erl
-module(step2_eval).

-include("types.hrl").

-export([main/1]).

main(_) ->
    ReplEnv = #{"+" => fun([A,B]) -> num(num(A) + num(B)) end,
                "-" => fun([A,B]) -> num(num(A) - num(B)) end,
                "*" => fun([A,B]) -> num(num(A) * num(B)) end,
                "/" => fun([A,B]) -> num(trunc(num(A) / num(B))) end},

    repl(ReplEnv).

num(#mal_num{val=V}) -> V;
num(V) when is_integer(V) -> #mal_num{val=V}.

repl(Env) ->
    case io:get_line(standard_io, "user> ") of
        eof ->
            io:format("~n"),
            ok;
        {error, ErrorDescription} ->
            io:format("Error reading input: ~p~n", [ErrorDescription]),
            exit(ioerr);
        Input ->
            case read(Input) of
                {success, Ast} ->
                    try eval(Ast, Env)
                             of {Res, NextEnv} -> io:format("~s~n", [print(Res)]),
                                                  repl(NextEnv)
                    catch
                        _:{malerr, Err} -> io:format("Exception: ~s~n", [Err]),
                                         repl(Env);
                        _:Reason -> io:format("Runtime error: ~p~n", [Reason]),
                                        repl(Env)
                    end;
                {error, Err} ->
                    io:format("Reader error: ~s~n", [Err]),
                    repl(Env)
            end
        end.

read(Code) ->
    reader:read_str(Code).

print(Ast) ->
    printer:pr_str(Ast).

eval([], Env) -> {[], Env};
eval(Ast, Env) when not is_record(Ast, mal_list) -> eval_ast(Ast, Env);
eval(Ast, Env) ->
    {#mal_list{elems=[F|Args]}, NextEnv} = eval_ast(Ast, Env),
%    io:format("> applying ~p with ~p~n", [F, Args]),
    {F(Args), NextEnv}.

eval_list_elem(Ast, {List, Env}) ->
    {Res, NextEnv} = eval(Ast, Env),
    {List ++ [Res], NextEnv}.

eval_ast(Ast, Env) when is_record(Ast, mal_list) ->
    {E, NextEnv} = lists:foldl(fun eval_list_elem/2, {[], Env}, Ast#mal_list.elems),
    {#mal_list{elems=E}, NextEnv};
eval_ast(Ast, Env) when is_record(Ast, mal_sym) ->
    try map_get(Ast#mal_sym.val, Env) of V -> {V, Env}
    catch
        _:{badkey, K} -> throw({malerr, io_lib:format("Symbol ~s not in env", [K])})
    end;
eval_ast(Ast, Env) -> {Ast, Env}.

File: impls/ts/step2_eval.ts
import { MalNumber, MalList, MalEnv, MalType, mal, MalFunc } from './types.ts'
import { read_str } from './reader.ts'
import { pr_str } from './printer.ts'

function malCalc(f: (x: number, y: number) => number): MalFunc {
    return mal.func((a: MalType, b: MalType) => {
        const x = a as MalNumber
        const y = b as MalNumber
        return mal.num(f(x.value, y.value))
    })
}

const env: MalEnv = {
    '+': malCalc((a,b) => a + b),
    '-': malCalc((a,b) => a - b),
    '*': malCalc((a,b) => a * b),
    '/': malCalc((a,b) => a / b),
}

function eval_ast(ast: MalType, env: MalEnv): MalType {
    switch(ast.type) {
        case 'symbol': {
            if(ast.value in env)
                return env[ast.value]
            else
                throw `Cannot find symbol '${ast.value}' in environment`
        }
        case 'list': {
            return ast.values.reduce(
                (a: MalList, b: MalType) => mal.list(a.values.concat([EVAL(b, env)])),
                mal.list([])
            );
        }
        default:
            return ast
    }
}

function READ(s: string): MalType {
    return read_str(s)
}
function EVAL(v: MalType, env: MalEnv): MalType {
    if(v.type != 'list') {
        return eval_ast(v, env)
    }
    else if(v.values.length === 0) {
        return v
    }
    else {
        const ml   = eval_ast(v, env) as MalList
        const f    = ml.values[0] as MalFunc
        const args = ml.values.slice(1)
        return f.value.apply(null, args)
    }
}
function PRINT(v: MalType): string {
    return pr_str(v)
}
function rep(s: string): string {
    return PRINT(EVAL(READ(s), env))
}

let line = prompt('user>')
while (line !== null) {
    try {
        console.log(rep(line));
    }
    catch(e) {
        console.log('Error: ', e)
    }
    line = prompt('user>')
}